Merge functions.md and lambdas.md design documents (#7425)
carbon-language/carbon-lang: docs/design/functions.md

<!-- tocstop -->

## Syntax

Functions are the core building block for applications. Carbon supports both
named function declarations and anonymous function expressions called _lambdas_.

A named function definition or declaration has one of the following syntactic
forms (where items in square brackets are optional and independent):

-   `fn` _name_ [_implicit-parameters_] [_tuple-pattern_] `=>` _expression_ `;`
-   `fn` _name_ [_implicit-parameters_] [_tuple-pattern_] [`->` _return-form_] `{`
    _statements_ `}`

> -   `fn` _name_ [_implicit-parameters_] [_tuple-pattern_] [`->` _return-form_] `;`

Reviewer: I think the tuple pattern is mandatory here, because a forward-declared function can't use positional parameters.

```suggestion
-   `fn` _name_ [_implicit-parameters_] _tuple-pattern_ [`->` _return-form_] `;`
```

A lambda expression has one of the following syntactic forms:

-   `fn` [_implicit-parameters_] [_tuple-pattern_] `=>` _expression_
-   `fn` [_implicit-parameters_] [_tuple-pattern_] [`->` _return-form_] `{` _statements_

>     `}`

Reviewer: Maybe instead we could mark the name as optional for the first and second function syntaxes, and then just say that a lambda is a function definition that omits the name? That way the reader doesn't have to cross-check to see if there are other differences between the two.

Reviewer 2: They're not quite the same. A function definition using `=>` has a trailing `;` but a lambda using `=>` does not.

> **TODO:** Update this section to cover the requirements on the form of the
> expression.

## Positional parameters

Positional parameters, denoted by a dollar sign followed by a non-negative
integer (for example, `$3`), are auto-typed parameters defined within the
function or lambda body when the explicit parameter list (parentheses) is
omitted. They are variadic by design, meaning an unbounded number of arguments
can be passed to any function or lambda that lacks an explicit parameter list.
Only the parameters that are named in the body will be read from, meaning the
highest named parameter denotes the minimum number of arguments required by the
function. The body is free to omit lower-numbered parameters (for example,
`fn { Print($10); }`).

This syntax was inspired by Swift's
[Shorthand Argument Names](https://docs.swift.org/swift-book/documentation/the-swift-programming-language/closures/#Shorthand-Argument-Names).

```carbon
let lambda: auto = fn => $0
```

```carbon
// A lambda that takes two positional parameters being used as a comparator
Sort(my_list, fn => $0.val < $1.val);
// In Swift: { $0.val < $1.val }
```

> 

Reviewer: I think we need to explain how positional parameters in nested functions work, and not just leave it implicit in the restriction below.

```suggestion

When positional parameters are used in a nested function definition, exactly one
of the enclosing functions must omit the explicit parameter list, and they are
interpreted as parameters of that function:

```carbon
fn F {
  fn G[let]() -> auto {
    // $0 is a parameter of F that is captured by G, not a parameter of G.
    return $0;
  }
}
```
```

### Positional parameter restrictions

Lambdas and named function definitions with positional parameters have the
following restrictions:

> -   They can only be used with function definitions, not forward declarations.

Reviewer: I think we should capture this in the grammar instead (see above).

Author: I made the change above, but I don't see harm repeating it here.

-   A single function or lambda cannot have both named and positional
    parameters. For example, if an explicit parameter list is provided

>     (parentheses), positional parameters cannot be used:

Reviewer: I think this is fully subsumed by the restriction on nesting, so I'm not sure we need it.

Author: Hopefully now addressed

    ```carbon
    fn Foo() {
      // In a variable:
      let lambda: auto = fn (v: auto) { Print(v); };
      // Equivalent in C++23:
      // const auto lambda = [](v: auto) -> void { Print(v); };

      // As an argument to a function call:
      Foo(fn (v: auto) { Print(v); });
      // Equivalent in C++23:
      // Foo([](v: auto) { Print(v); });
    }

>     ```

Reviewer: This example doesn't seem very on-point. It should probably instead be an example of code that violates the restriction, with comments marking it as incorrect.

Author: Fixed.

    However, deduced parameters in the implicit parameter list are always
    permitted:

    ```carbon
    fn Foo() {
      let lambda: auto = fn [T:! Printable](t: T) { Print(t); };
    }
    ```

-   Positional parameters can only be used in a context where there is exactly
    one enclosing function or lambda that has no explicit parameter list. For

>     example:

Reviewer: As noted above, I think we need to explain how positional parameters work in nested functions. That explanation naturally incorporates this restriction, so we can probably move some of these examples up there, and then omit this point.

Author: I rephrased it as an example consequence of the rule.

        fn => $0 < $1
      );
    }

    fn Foo4() {
      let lambda: auto = fn -> bool {
        // ❌ Invalid: Outer lambda is already using positional parameters
        return (fn => $0 < $1)($0, $1);
      };
    }

    fn Foo5() {
      let lambda: auto = fn (x: i32, y: i32) -> bool {
        // ✅ Valid: Outer lambda has explicit parameters
        return (fn => $0 < $1)(x, y);
      };
    }
    ```

## Captures

Captures in Carbon mirror the non-init captures of C++. A capture declaration
consists of a capture mode (for `var` captures) followed by the name of a
binding from the enclosing scope, and makes that identifier available in the
inner function body. The lifetime of a capture is the lifetime of the function
in which it exists. For example:

```carbon
fn Foo() {
  let handle: Handle = Handle.Get();
  var thread: Thread = Thread.Make(fn [var handle] { handle.Process(); });
  thread.Join();
}
```

```carbon
fn Foo() {
  let handle: Handle = Handle.Get();
  fn MyThread[handle]() { handle.Process(); }
  var thread: Thread = Thread.Make(MyThread);
  thread.Join();
}
```

For both named functions and lambdas, these captures (and function fields) are

> specified in square brackets `[`...`]` as part of the implicit parameter list.

Reviewer: I think this should be folded into the first paragraph (without the mention of function fields, which I think we should just omit). It also doesn't need a separate example, because the examples above already illustrate it (which is part of why it needs to be moved up).

Author: I've rearranged things but I've kept both examples, since I think one was specifically supposed to show the lifetime issues.

Reviewer: The `Handle` example was originally the only example in this section, so I think it was meant to show captures overall, not just their lifetime properties. Among other things, I don't think it illustrates the lifetime issues very clearly, because it would work correctly under pretty much any plausible lifetime rules (maybe we should drop the `Join` calls?).

The `Print` example seems to have been taken from a completely different section, "Implicit parameters in square brackets", and I don't think it makes sense here, since it makes extensive use of function fields (which haven't been introduced yet). If you remove the function fields, it might make sense as a simpler example that doesn't implicate lifetime, but I think the `Handle` example is simple enough to be the first example, even with the additional complication of lifetimes.

Author: I deleted the print example, and created some space between the handle example and the statement about lifetimes.

For example:

```carbon
fn Foo(x: i32) {
  // In a variable:
  let lambda: auto = fn [var x, var y: i32 = 0] { Print(++x, ++y); };
  // Equivalent in C++23:
  // const auto lambda = [x, y = int32_t{0}] mutable -> void { Print(++x, ++y); };

  // As an argument to a function call:
  Foo(fn [var x, var y: i32 = 0] { Print(++x, ++y); });
  // Equivalent in C++23:
  // Foo([x, y = int32_t{0}] mutable -> void { Print(++x, ++y); });
}
```

### Capture restrictions on named functions

While lambdas can use captures freely, named function definitions support
captures (and function fields) with these restrictions:

-   They can only be used on functions where the definition is attached to the
    declaration (so they cannot be forward declared).
-   Captures and function fields are only supported on local function
    definitions immediately defined inside the body of another function. They

>     are not supported on member functions of classes/interfaces.

Reviewer: Calling out classes and interfaces makes out sound like, by omission, they _are_ supported on package- and impl-scope functions. The simplest fix is probably to just drop this sentence, since it's implied by the previous sentence.

```suggestion
    definitions immediately defined inside the body of another function.
```

expression. We need a mutable handle to the function in order to be able to
mutate its mutable state.

### Default capture mode

By default, there is no capturing in lambdas and functions. The lack of any
square brackets is the same as an empty pair of square brackets. Users can opt
into capturing behavior. This is done either by way of individual explicit
captures, or more succinctly by way of a default capture mode. The default
capture mode roughly mirrors the syntax `[=]` and `[&]` capture modes from C++
by being the first thing to appear in the square brackets.

```carbon
fn Foo1() {
  let handle: Handle = Handle.Get();
  fn MyThread[var]() {
    // `handle` is captured by-object due to the default capture
    // mode specifier of `var`
    handle.Process();
  }
  var thread: Thread = Thread.Make(MyThread);
  thread.Join();
}

fn Foo2() {
  let handle: Handle = Handle.Get();
  fn MyThread[let]() {
    // `handle` is captured by-value due to the default capture
    // mode specifier of `let`
    handle.Process();
  }
  var thread: Thread = Thread.Make(MyThread);
  thread.Join();
}
```

### Function fields

> Function fields mirror the behavior of init captures in C++. A function field

Reviewer: ```suggestion
Function fields mirror the behavior of init captures in C++. Function fields
are defined in the implicit parameter list, and are allowed only where
captures are allowed. A function field
```

This way we don't need to mention function fields in the earlier sections, before we've explained what they are.

also applies to captures.

The other case is by-value function fields. Since C++ const references, when
made into fields of a class, prevent the class from being copy assigned, so too
should by-value function fields prevent the function in which it is contained
from being copy assigned.

## Lambdas

One goal of Carbon's lambda syntax is to have continuity between lambdas and
named functions. Below are some example declarations:

Implicit return types:

```carbon
// In a variable:
let lambda: auto = fn => T.Make();
// Equivalent in C++23:
// const auto lambda = [] { return T::Make(); };

// As an argument to a function call:
Foo(10, 20, fn => T.Make());
// Equivalent in C++23:
// Foo(10, 20, [] { return T::Make(); });
```

Explicit return types:

```carbon
// In a variable:
let lambda: auto = fn -> T { return T.Make(); };
// Equivalent in C++23:
// const auto lambda = [] -> T { return T::Make(); };

// As an argument to a function call:
PushBack(my_list, fn -> T { return T.Make() });
// Equivalent in C++23:
// PushBack(my_list, [] { return T::Make(); });
```

### Lambdas may not take `self` as a parameter

To mirror C++'s use of capturing `this`, `self` should always come from the

> outer scope as a capture. `self: Self` is never permitted on lambdas.

Reviewer: I think the phrasing needs to be tweaked to reflect #7016.

```suggestion
outer scope as a capture. `self` is never permitted in the explicit
parameter list of a lambda.
```
